perma-payments/perma_payments/views.py

```diff
@@ -0,0 +1,428 @@
+@require_http_methods(["POST"])
+@sensitive_post_parameters(*SENSITIVE_POST_PARAMETERS)
+def cybersource_callback(request):
+    """
+    In dev, curl http://192.168.99.100/cybersource-callback/ -X POST -d '@/Users/rcremona/code/perma-payments/sample_response.txt'
+    """
+    try:
+        data = process_cybersource_transmission(request.POST, FIELDS_REQUIRED_FROM_CYBERSOURCE['cybersource_callback'])
+    except InvalidTransmissionException:
+        return bad_request(request)
+
+    related_request = OutgoingTransaction.objects.get(transaction_uuid=data['req_transaction_uuid'])
+    decision = data['decision']
+    reason_code = data['reason_code']
+    message = data['message']
+
+    if isinstance(related_request, UpdateRequest):
+        Response.save_new_with_encrypted_full_response(
+            UpdateRequestResponse,
+            request.POST,
+            {
+                'related_request': related_request,
+                'decision': decision,
+                'reason_code': reason_code,
+                'message': message,
+            }
+        )
+
+    elif isinstance(related_request, SubscriptionRequest):
+        payment_token = request.POST.get('payment_token', '')
+
+        # Perma-Payments does not support 16-digit format-preserving Payment Tokens.
+        # See docstring for SubscriptionAgreement model for details.
+        if len(payment_token) == 16:
+            logger.error("16-digit Payment Token received in response to subscription request {}. Not supported by Perma-Payments! Investigate ASAP.".format(related_request.pk))
+
+        Response.save_new_with_encrypted_full_response(
+            SubscriptionRequestResponse,
+            request.POST,
+            {
+                'related_request': related_request,
+                'decision': decision,
+                'reason_code': reason_code,
+                'message': message,
+                'payment_token': payment_token,
+            }
+        )
+        related_request.subscription_agreement.update_status_after_cs_decision(decision, {k: v for (k, v) in request.POST.items() if k not in SENSITIVE_POST_PARAMETERS})
+
+    else:
+        raise NotImplementedError("Can't handle a response of type {}, returned in reponse to outgoing transaction {}".format(type(related_request), related_request.pk))
+
+    return render(request, 'generic.html', {'heading': 'CyberSource Callback', 'message': 'OK'})
+
+
+@csrf_exempt
+@require_http_methods(["POST"])
+def subscription(request):
+    """
+    Returns a simplified version of a registrar's subscription status,
+    as needed for making decisions in Perma.
+    """
+    try:
+        data = process_perma_transmission(request.POST, FIELDS_REQUIRED_FROM_PERMA['subscription'])
+    except InvalidTransmissionException:
+        return bad_request(request)
+
+    standing_subscription = SubscriptionAgreement.registrar_standing_subscription(data['registrar'])
+    if not standing_subscription:
+        subscription = None
```

Comment:
shadows subscription method name, should be changed, right?

Reply:
So I think this one (unlike the import situation) is actually fine and isn't really shadowing: the inside of the function has a different namespace. So, setting subscription = None doesn't overwrite anything, it makes a new variable, in the function's own namespace, called subscription. I don't _think_ that's a stylistic problem, and am tempted to hang onto it since the variable name so clearly describes the object, but I could be talked out of it. What do you think @jcushman? Is it poor style to take advantage of python namespacing to reuse a convenient name?